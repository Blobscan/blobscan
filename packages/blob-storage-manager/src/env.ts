import {
  z,
  booleanSchema,
  createEnv,
  presetEnvOptions,
  maskSensitiveData,
} from "@blobscan/zod";

export function parseEnv() {
  return createEnv({
    envOptions: {
      server: {
        BEE_DEBUG_ENDPOINT: z.string().url().optional(),
        BEE_ENDPOINT: z.string().url().optional(),
        CHAIN_ID: z.coerce.number().positive().default(1),
        FILE_SYSTEM_STORAGE_ENABLED: booleanSchema.default("false"),
        FILE_SYSTEM_STORAGE_PATH: z.string().optional(),
        GOOGLE_STORAGE_BUCKET_NAME: z.string().optional(),
        GOOGLE_STORAGE_PROJECT_ID: z.string().optional(),
        GOOGLE_SERVICE_KEY: z.string().optional(),
        GOOGLE_STORAGE_API_ENDPOINT: z.string().optional(),
        GOOGLE_STORAGE_ENABLED: booleanSchema.default("false"),
        POSTGRES_STORAGE_ENABLED: booleanSchema.default("true"),
        SWARM_STORAGE_ENABLED: booleanSchema.default("false"),
      },

      ...presetEnvOptions,
    },
    display(env) {
      console.log(
        `Blob storage manager configuration: chainId=${env.CHAIN_ID}, file_system=${env.FILE_SYSTEM_STORAGE_ENABLED} postgres=${env.POSTGRES_STORAGE_ENABLED}, gcs=${env.GOOGLE_STORAGE_ENABLED}, swarm=${env.SWARM_STORAGE_ENABLED}`
      );

      if (env.GOOGLE_STORAGE_ENABLED) {
        console.log(
          `GCS configuration: bucketName=${
            env.GOOGLE_STORAGE_BUCKET_NAME
          }, projectId=${maskSensitiveData(
            env.GOOGLE_STORAGE_PROJECT_ID
          )}, apiEndpoint=${env.GOOGLE_STORAGE_API_ENDPOINT}`
        );
      }

      if (env.SWARM_STORAGE_ENABLED) {
        console.log(
          `Swarm configuration: beeEndpoint=${env.BEE_ENDPOINT}, debugEndpoint=${env.BEE_DEBUG_ENDPOINT}`
        );
      }

      if (env.FILE_SYSTEM_STORAGE_ENABLED) {
        console.log(
          `File system configuration: blobDirPath=${env.FILE_SYSTEM_STORAGE_PATH}`
        );
      }
    },
  });
}
export const env = parseEnv();

export type Environment = typeof env;
