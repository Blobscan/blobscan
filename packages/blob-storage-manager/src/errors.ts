class ErrorException extends Error {
  constructor(message: string, cause?: Error | Error[]) {
    super(message, {
      cause,
    });

    this.name = this.constructor.name;
  }
}

export class BlobStorageError extends ErrorException {
  constructor(storageName: string, message: string, cause?: Error) {
    super(`${storageName} failed: ${message}`, cause);
  }
}

export class StorageCreationError extends BlobStorageError {
  constructor(storageName: string, message: string, cause: Error) {
    super(
      storageName,
      `Creation failed${message.length ? `: ${message}` : ""}`,
      cause
    );
  }
}

export class BlobStorageManagerError extends ErrorException {
  constructor(message: string, cause?: Error | BlobStorageError[]) {
    super(message, cause);
  }
}
