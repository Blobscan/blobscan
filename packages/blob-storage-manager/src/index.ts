export { BlobStorageManager } from "./BlobStorageManager";
export * from "./blob-storage-manager";
export { env } from "./env";
export type { Environment } from "./env";
export * from "./errors";
export { GoogleStorage, PostgresStorage, SwarmStorage } from "./storages";
export * from "./types";
