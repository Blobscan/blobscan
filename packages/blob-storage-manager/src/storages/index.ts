export * from "./FileSystemStorage";
export * from "./GoogleStorage";
export * from "./PostgresStorage";
export * from "./SwarmStorage";
