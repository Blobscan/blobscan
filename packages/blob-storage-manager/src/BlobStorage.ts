import type { Environment } from "./env";
import { BlobStorageError, StorageCreationError } from "./errors";
import type { BlobStorageName } from "./types";

export interface BlobStorageConfig {
  chainId: number;
}

export abstract class BlobStorage {
  chainId: number;
  name: BlobStorageName;

  constructor(name: BlobStorageName, chainId: number) {
    this.name = name;
    this.chainId = chainId;
  }

  protected abstract _healthCheck(): Promise<void>;
  protected abstract _getBlob(uri: string): Promise<string>;
  protected abstract _storeBlob(hash: string, data: string): Promise<string>;

  async healthCheck(): Promise<"OK"> {
    try {
      await this._healthCheck();

      return "OK";
    } catch (err) {
      throw new BlobStorageError(
        this.constructor.name,
        "Storage is not reachable",
        err as Error
      );
    }
  }

  async getBlob(uri: string): Promise<string> {
    try {
      const blob = await this._getBlob(uri);

      return blob;
    } catch (err) {
      throw new BlobStorageError(
        this.constructor.name,
        `Failed to get blob with uri "${uri}"`,
        err as Error
      );
    }
  }

  async storeBlob(hash: string, data: string): Promise<string> {
    try {
      const res = await this._storeBlob(hash, data);

      return res;
    } catch (err) {
      throw new BlobStorageError(
        this.constructor.name,
        `Failed to store blob with hash "${hash}"`,
        err as Error
      );
    }
  }

  protected buildBlobFileName(hash: string): string {
    return `${this.chainId.toString()}/${hash.slice(2, 4)}/${hash.slice(
      4,
      6
    )}/${hash.slice(6, 8)}/${hash.slice(2)}.txt`;
  }

  static async create<T extends BlobStorage, C extends BlobStorageConfig>(
    this: new (config: C) => T,
    config: C
  ): Promise<T> {
    try {
      const blobStorage = new this(config);

      await blobStorage.healthCheck();

      return blobStorage;
    } catch (err) {
      const err_ = err as Error;
      throw new StorageCreationError(
        this.name,
        err_.message,
        err_.cause as Error
      );
    }
  }

  static async tryCreateFromEnv<
    C extends BlobStorageConfig,
    T extends BlobStorage
  >(
    this: {
      new (config: C): T;
      getConfigFromEnv(env: Partial<Environment>): C;
    },
    env: Partial<Environment>
  ): Promise<[T?, StorageCreationError?]> {
    let config: C;

    try {
      config = this.getConfigFromEnv(env);
    } catch (err) {
      const creationError = new StorageCreationError(
        this.name,
        "Failed to get config",
        err as Error
      );
      return [, creationError];
    }

    try {
      const blobStorage = new this(config);

      await blobStorage.healthCheck();

      return [blobStorage];
    } catch (err) {
      const err_ = err as Error;
      const creationError = new StorageCreationError(
        this.name,
        err_.message,
        err_.cause as Error
      );
      return [, creationError];
    }
  }

  protected static getConfigFromEnv(
    env: Partial<Environment>
  ): BlobStorageConfig {
    if (!env.CHAIN_ID) {
      throw new BlobStorageError(
        this.name,
        `No config variables found: no chain id provided`
      );
    }

    return {
      chainId: env.CHAIN_ID,
    };
  }
}
