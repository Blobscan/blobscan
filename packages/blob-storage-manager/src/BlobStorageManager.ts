import { api, SemanticAttributes } from "@blobscan/open-telemetry";

import type { BlobStorage } from "./BlobStorage";
import { BlobStorageManagerError } from "./errors";
import type { BlobStorageError } from "./errors";
import { tracer, updateBlobStorageMetrics } from "./instrumentation";
import type {
  Blob,
  BlobReference,
  BlobStorageName,
  StorageOf,
  StoreOptions,
} from "./types";
import { calculateBlobBytes, removeDuplicatedStorages } from "./utils";

export class BlobStorageManager {
  #blobStorages: BlobStorage[];

  constructor(blobStorages: BlobStorage[]) {
    if (!blobStorages.length) {
      throw new BlobStorageManagerError("No blob storages provided");
    }

    this.#blobStorages = removeDuplicatedStorages(blobStorages);
  }

  getStorage<T extends BlobStorageName>(name: T): StorageOf<T> | undefined {
    const blobStorage = this.#blobStorages.find(
      (storage): storage is StorageOf<T> => storage.name === name
    );

    return blobStorage as StorageOf<T> | undefined;
  }

  hasStorage<T extends BlobStorageName>(name: T): boolean {
    return !!this.getStorage(name);
  }

  async getBlob(
    ...blobReferences: BlobReference<BlobStorageName>[]
  ): Promise<{ data: string; storage: BlobStorageName } | null> {
    return tracer.startActiveSpan(
      "blob_storage_manager",
      {
        attributes: {
          [SemanticAttributes.CODE_FUNCTION]: "getBlob",
        },
      },
      async (span) => {
        const availableReferences = blobReferences.filter(({ storage }) =>
          this.hasStorage(storage)
        );

        try {
          const blob = await Promise.any(
            availableReferences.map(({ reference, storage: storageName }) =>
              tracer.startActiveSpan(
                "blob_storage_manager:storage",
                {
                  attributes: {
                    storage: storageName,
                    [SemanticAttributes.CODE_FUNCTION]: "getBlob",
                  },
                },
                async (storageSpan) => {
                  const start = performance.now();
                  const storage = this.getStorage(storageName);

                  if (!storage) {
                    throw new BlobStorageManagerError(
                      `Storage "${storageName}" is not available`
                    );
                  }

                  const blob = await storage
                    .getBlob(reference)
                    .then((data) => ({
                      data,
                      storage: storageName,
                    }));
                  const end = performance.now();

                  storageSpan.end();

                  updateBlobStorageMetrics({
                    storage: storageName,
                    blobSize: calculateBlobBytes(blob.data),
                    direction: "received",
                    duration: end - start,
                  });

                  return blob;
                }
              )
            )
          );

          span.end();

          return blob;
        } catch (e) {
          const message = "Failed to get blob from any of the storages";
          let err: BlobStorageManagerError;

          // Aggregate errors are thrown when all the promises fail
          if (e instanceof AggregateError) {
            err = new BlobStorageManagerError(
              message,
              e.errors as BlobStorageError[]
            );
          } else {
            err = new BlobStorageManagerError(message, e as Error);
          }

          span.setStatus({ code: api.SpanStatusCode.ERROR });
          span.recordException(err);

          throw err;
        } finally {
          span.end();
        }
      }
    );
  }

  async storeBlob(
    { data, versionedHash }: Blob,
    opts?: StoreOptions
  ): Promise<{
    references: BlobReference<BlobStorageName>[];
    errors: BlobStorageError[];
  }> {
    return tracer.startActiveSpan(
      "blob_storage_manager",
      {
        attributes: {
          [SemanticAttributes.CODE_FUNCTION]: "storeBlob",
        },
      },
      async (span) => {
        const selectedStorages = this.#getSelectedStorages(opts);

        const results = await Promise.allSettled(
          selectedStorages.map((storage) => {
            return tracer.startActiveSpan(
              "blob_storage_manager:storage",
              {
                attributes: {
                  storage: storage.name,
                  [SemanticAttributes.CODE_FUNCTION]: "storeBlob",
                },
              },
              async (storageSpan) => {
                const start = performance.now();
                const blobReference = await storage
                  .storeBlob(versionedHash, data)
                  .then<BlobReference<BlobStorageName>>((reference) => ({
                    reference,
                    storage: storage.name,
                  }));
                const end = performance.now();

                storageSpan.end();

                updateBlobStorageMetrics({
                  blobSize: calculateBlobBytes(data),
                  direction: "sent",
                  duration: end - start,
                  storage: storage.name,
                });

                return blobReference;
              }
            );
          })
        );
        const references = results
          .filter(
            (
              res
            ): res is PromiseFulfilledResult<BlobReference<BlobStorageName>> =>
              res.status === "fulfilled"
          )
          .map((res) => res.value);
        const storageErrors = results
          .filter((r) => r.status === "rejected")
          .map((r) => (r as PromiseRejectedResult).reason as BlobStorageError);

        if (!references.length) {
          const err = new BlobStorageManagerError(
            `Failed to upload blob with hash "${versionedHash}" to any storage`,
            storageErrors
          );

          span.setStatus({ code: api.SpanStatusCode.ERROR });
          span.recordException(err);
          throw err;
        }

        span.end();

        return {
          references,
          errors: storageErrors,
        };
      }
    );
  }

  #getSelectedStorages(
    { selectedStorages }: StoreOptions = { selectedStorages: [] }
  ): BlobStorage[] {
    const uniqueInputStorageNames = Array.from(new Set(selectedStorages));
    const selectedAvailableStorages = uniqueInputStorageNames.length
      ? this.#blobStorages.filter((s) =>
          uniqueInputStorageNames.includes(s.name)
        )
      : // Use all available storages if none are selected
        this.#blobStorages;

    if (
      uniqueInputStorageNames.length &&
      selectedAvailableStorages.length !== uniqueInputStorageNames.length
    ) {
      const selectdAvailableStorageNames = selectedAvailableStorages.map(
        (s) => s.name
      );
      const missingStorageNames = uniqueInputStorageNames.filter(
        (storageName) => !selectdAvailableStorageNames.includes(storageName)
      );

      throw new BlobStorageManagerError(
        `Some of the selected storages are not available: ${missingStorageNames.join(
          ", "
        )}"`
      );
    }

    return selectedAvailableStorages;
  }
}
