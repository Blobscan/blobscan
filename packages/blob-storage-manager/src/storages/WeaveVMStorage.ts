import { BlobStorage as BlobStorageName } from "@blobscan/db/prisma/enums";
import { z } from "@blobscan/zod";

import type { BlobStorageConfig } from "../BlobStorage";
import { BlobStorage } from "../BlobStorage";

const blobResponseSchema = z.object({
  blob_data: z.string(),
});

export interface WeaveVMStorageConfig extends BlobStorageConfig {
  apiBaseUrl: string;
}

export class WeaveVMStorage extends BlobStorage {
  apiBaseUrl: string;

  protected constructor({ apiBaseUrl, chainId }: WeaveVMStorageConfig) {
    super(BlobStorageName.WEAVEVM, chainId);

    this.apiBaseUrl = apiBaseUrl;
  }

  protected async _healthCheck(): Promise<void> {
    await fetch(`${this.apiBaseUrl}/v1/stats`);
  }

  protected async _getBlob(uri: string): Promise<string> {
    const response = await fetch(`${this.apiBaseUrl}/v1/blob/${uri}`);
    const jsonRes = await response.json();

    const res = blobResponseSchema.safeParse(jsonRes);

    if (!res.success) {
      throw new Error("Failed to parse blob response", {
        cause: res.error.flatten().fieldErrors.blob_data?.join(";"),
      });
    }

    return res.data.blob_data;
  }

  protected async _storeBlob(_: string, __: Buffer): Promise<string> {
    throw new Error(
      '"_storeBlob" operation is not allowed for WeaveVM storage'
    );
  }

  protected _stageBlob(_: string, __: Buffer): Promise<string> {
    throw new Error(
      '"_stageBlob" operation is not allowed for WeaveVM storage'
    );
  }

  protected async _removeBlob(_: string): Promise<void> {
    throw new Error(
      '"_removeBlob" operation is not allowed for WeaveVM storage'
    );
  }

  getBlobUri(hash: string) {
    return hash;
  }

  static async create(config: WeaveVMStorageConfig) {
    try {
      const storage = new WeaveVMStorage(config);

      await storage.healthCheck();

      return storage;
    } catch (err) {
      const err_ = err as Error;

      throw new Error("Failed to create WeaveVM storage", {
        cause: err_,
      });
    }
  }
}
