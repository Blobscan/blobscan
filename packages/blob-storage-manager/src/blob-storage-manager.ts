import { logger } from "@blobscan/logger";

import type { BlobStorage } from "./BlobStorage";
import { BlobStorageManager } from "./BlobStorageManager";
import { env } from "./env";
import type { Environment } from "./env";
import type { BlobStorageName } from "./types";
import { BLOB_STORAGE_NAMES, createStorageFromEnv } from "./utils";

let blobStorageManager: BlobStorageManager | undefined;

function isBlobStorageEnabled(storageName: BlobStorageName) {
  const storageEnabledKey =
    `${storageName}_STORAGE_ENABLED` as keyof Environment;
  const storageEnabled = env[storageEnabledKey];

  return storageEnabled === true || storageEnabled === "true";
}

async function createBlobStorageManager() {
  const blobStorages = await Promise.all(
    Object.values(BLOB_STORAGE_NAMES).map(async (storageName) => {
      if (isBlobStorageEnabled(storageName)) {
        const [storage, storageError] = await createStorageFromEnv(storageName);

        if (storageError) {
          logger.warn(
            `${storageError.message}. Caused by: ${storageError.cause}`
          );
        }

        return storage;
      }
    })
  );

  const availableStorages = blobStorages.filter(
    (storage): storage is BlobStorage => !!storage
  );

  return new BlobStorageManager(availableStorages);
}

export async function getBlobStorageManager(): Promise<BlobStorageManager> {
  if (!blobStorageManager) {
    blobStorageManager = await createBlobStorageManager();
  }

  return blobStorageManager;
}
