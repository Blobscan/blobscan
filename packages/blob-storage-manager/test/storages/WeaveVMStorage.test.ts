import { http, HttpResponse } from "msw";
import type { SetupServerApi } from "msw/node";
import { setupServer } from "msw/node";
import { afterEach, beforeAll, beforeEach, describe, expect, it } from "vitest";

import { env } from "@blobscan/env";
import { testValidError } from "@blobscan/test";

import { BlobStorageError } from "../../src/errors";
import { WeaveVMStorage } from "../../src/storages/WeaveVMStorage";

const MOCK_API_BASE_URL = "https://blobscan.weavevm";

const MOCK_RESPONSES: Record<string, string> = {
  "0x01b4b4b2b62f7a206efcb78e2ed6ef5bde9c1ac33d4d44eccfd43aa951ef1d22":
    "0x4fe40fc67f9c3a3ffa2be77d10fe7818",
};

class WeaveVMStorageMock extends WeaveVMStorage {
  constructor() {
    super({
      apiBaseUrl: MOCK_API_BASE_URL,
      chainId: env.CHAIN_ID,
    });
  }

  healthCheck() {
    return super.healthCheck();
  }
}

describe("WeavevmStorage", () => {
  let weavevmServer: SetupServerApi;
  let storage: WeaveVMStorageMock;

  beforeAll(() => {
    weavevmServer = setupServer(
      http.get(`${MOCK_API_BASE_URL}/v1/blob/:uri`, ({ request }) => {
        const uri = request.url.split("/").pop();

        if (!uri) {
          return HttpResponse.json({
            code: 400,
            message: "invalid path params",
          });
        }

        const blobData = MOCK_RESPONSES[uri];

        if (!blobData) {
          return HttpResponse.json({
            code: 404,
            message: "blob not found",
          });
        }

        return HttpResponse.json({
          blob_data: blobData,
        });
      }),
      http.get(`${MOCK_API_BASE_URL}/v1/stats`, () => {
        return HttpResponse.json({
          blob_versioned_hash:
            "0x01b4b4b2b62f7a206efcb78e2ed6ef5bde9c1ac33d4d44eccfd43aa951ef1d22",
          last_archived_eth_block: 19559986,
          wvm_archive_txid:
            "0x01ee8325bc5607a16dd64ff2bcbec7d596b170f31def52615abf6b3f25ceb5a5",
        });
      })
    );

    weavevmServer.listen();

    return () => {
      weavevmServer.close();
    };
  });

  beforeEach(() => {
    storage = new WeaveVMStorageMock();
  });

  afterEach(() => {
    weavevmServer.resetHandlers();
  });

  it("should create a storage", async () => {
    const storage_ = await WeaveVMStorage.create({
      chainId: env.CHAIN_ID,
      apiBaseUrl: MOCK_API_BASE_URL,
    });

    expect(storage_.chainId, "Chain ID mismatch").toBe(env.CHAIN_ID);
    expect(storage_.apiBaseUrl).toBe(MOCK_API_BASE_URL);
  });

  it("return the correct uri given a blob hash", () => {
    const blobHash = "exampleBlobHash";
    const blobUri = storage.getBlobUri(blobHash);

    expect(blobUri).toBe(blobHash);
  });

  it("should return 'OK' if storage is healthy", async () => {
    await expect(storage.healthCheck()).resolves.toBe("OK");
  });

  it("should get a blob given its reference", async () => {
    const blobHash = Object.keys(MOCK_RESPONSES)[0] as string;

    const result = await storage.getBlob(blobHash);

    expect(result).toBe(MOCK_RESPONSES[blobHash]);
  });

  testValidError(
    "should fail when trying to parse an invalid blob retrieval response",
    async () => {
      weavevmServer.use(
        http.get(`${MOCK_API_BASE_URL}/v1/blob/:uri`, () => {
          return HttpResponse.json({
            blob_data: 123,
          });
        })
      );

      await storage.getBlob("invalidBlobHash");
    },
    BlobStorageError,
    {
      checkCause: true,
    }
  );

  testValidError(
    "should fail when trying to store a blob",
    async () => {
      await storage.storeBlob("blobHash", "blobData");
    },
    BlobStorageError,
    {
      checkCause: true,
    }
  );

  testValidError(
    "should fail when trying to remove a blob",
    async () => {
      await storage.removeBlob("blobHash");
    },
    BlobStorageError,
    {
      checkCause: true,
    }
  );
});
