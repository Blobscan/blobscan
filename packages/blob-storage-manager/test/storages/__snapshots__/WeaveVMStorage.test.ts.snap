// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WeavevmStorage > should fail when trying to parse an invalid blob retrieval response 1`] = `"WeaveVMStorageMock failed: Failed to get blob with uri \\"invalidBlobHash\\""`;

exports[`WeavevmStorage > should fail when trying to parse an invalid blob retrieval response 2`] = `[Error: Failed to parse blob response]`;

exports[`WeavevmStorage > should fail when trying to remove a blob 1`] = `"WeaveVMStorageMock failed: Failed to remove blob with uri \\"blobHash\\""`;

exports[`WeavevmStorage > should fail when trying to remove a blob 2`] = `[Error: "_removeBlob" operation is not allowed for WeaveVM storage]`;

exports[`WeavevmStorage > should fail when trying to store a blob 1`] = `"WeaveVMStorageMock failed: Failed to store blob with hash \\"0x01aaeb1f6ac28c7dc38569bced04f3c7d8b8d16ee2f1818b5b15b2004a5a04eb\\""`;

exports[`WeavevmStorage > should fail when trying to store a blob 2`] = `[Error: "_storeBlob" operation is not allowed for WeaveVM storage]`;
